src/derive/curve.rs

```diff
@@ -625,6 +664,71 @@ macro_rules! new_curve_impl {
             }
         }
 
+        impl group::Group for $name_jac_ext {
+            type Scalar = $scalar;
+
+            fn random(mut rng: impl RngCore) -> Self {
+                let affine = $name_affine::random(&mut rng);
+                $name_jac_ext::from(affine)
+            }
+
+            /// <http://www.hyperelliptic.org/EFD/g1p/auto-shortw-xyzz.html#doubling-dbl-2008-s-1>
+            #[inline]
+            fn double(&self) -> Self {
+
+                // p
+                let p = self;
+                let x1 = p.x;
+                let y1 = p.y;
+                let zz1 = p.zz;
+                let zzz1 = p.zzz;
+
+                // curve constants
+                let a = Self::curve_constant_a();
```

Comment:
> a is always equal to 0 for pairing curves (those needed for Snarks) and it's also equal to 0 for secp256k1.
> 
> If Rust provides compile time evaluation, it would be extremely useful to specialize for a == 0

I have changed it to:

    // curve constants
    const A: $base = $name_jac_ext::curve_constant_a();

This should enforce compile time evaluation and hopefully imply that

    let m = (x1_sqr.double()+x1_sqr) + A*zz1.square();
    
gets optimized to

    let m = (x1_sqr.double()+x1_sqr) 
    
for the curves where $a=0$.  But I am currently not sure how to verify this.

Reply:
Anyway, I tried commenting out the last term and then ran `cargo test`.  All tests pass.

I put in a `debug_assert_eq!(a, 0);`, so if we can catch future curves where `a != 0`.

Reply:
You should use if A == 0 and if A != 0.

The multiplication and addition operations are pure assembly and will not be optimized away.

Reply:
We can make it that way, no problem, but I politely remain skeptical of the argument.  When the value is known to be 0 at compile time, like in our case, the assembly will indeed be optimized away.  I tested it with `-opt-level=2` here https://godbolt.org/z/4Gr41o8Kr

Reply:
> We can make it that way, no problem, but I politely remain skeptical of the argument. When the value is known to be 0 at compile time, like in our case, the assembly will indeed be optimized away. I tested it with `-opt-level=2` here https://godbolt.org/z/4Gr41o8Kr

I realise that `*` is overloaded for field arithmetic, so my argument may or may not be applicable.